src/com/trollworks/gcs/spell/SpellTechniqueEditor.java

```diff
@@ -0,0 +1,415 @@
+/*
+ * Copyright (c) 1998-2020 by Richard A. Wilkes. All rights reserved.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public License, version 2.0.
+ * If a copy of the MPL was not distributed with this file, You can obtain one at
+ * http://mozilla.org/MPL/2.0/.
+ *
+ * This Source Code Form is "Incompatible With Secondary Licenses", as defined by the
+ * Mozilla Public License, version 2.0.
+ */
+
+package com.trollworks.gcs.spell;
+
+import com.trollworks.gcs.character.GURPSCharacter;
+import com.trollworks.gcs.prereq.PrereqsPanel;
+import com.trollworks.gcs.skill.SkillAttribute;
+import com.trollworks.gcs.skill.SkillDefault;
+import com.trollworks.gcs.skill.SkillDefaultType;
+import com.trollworks.gcs.skill.SkillDifficulty;
+import com.trollworks.gcs.skill.SkillLevel;
+import com.trollworks.gcs.skill.Technique;
+import com.trollworks.gcs.weapon.MeleeWeaponEditor;
+import com.trollworks.gcs.weapon.RangedWeaponEditor;
+import com.trollworks.gcs.weapon.WeaponStats;
+import com.trollworks.gcs.widgets.outline.ListRow;
+import com.trollworks.gcs.widgets.outline.RowEditor;
+import com.trollworks.toolkit.ui.UIUtilities;
+import com.trollworks.toolkit.ui.layout.ColumnLayout;
+import com.trollworks.toolkit.ui.widget.LinkedLabel;
+import com.trollworks.toolkit.ui.widget.outline.OutlineModel;
+import com.trollworks.toolkit.utility.I18n;
+import com.trollworks.toolkit.utility.text.NumberFilter;
+import com.trollworks.toolkit.utility.text.Numbers;
+import com.trollworks.toolkit.utility.text.Text;
+
+import java.awt.Component;
+import java.awt.Container;
+import java.awt.Dimension;
+import java.awt.event.ActionEvent;
+import java.awt.event.ActionListener;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
+import javax.swing.JCheckBox;
+import javax.swing.JComboBox;
+import javax.swing.JLabel;
+import javax.swing.JPanel;
+import javax.swing.JScrollPane;
+import javax.swing.JTabbedPane;
+import javax.swing.JTextField;
+import javax.swing.SwingConstants;
+import javax.swing.event.DocumentEvent;
+import javax.swing.event.DocumentListener;
+import javax.swing.text.Document;
+
+public class SpellTechniqueEditor extends RowEditor<SpellTechnique> implements ActionListener, DocumentListener {
```

Comment:
It appears that most of this file is copied from SpellEditor.java. Can a common base class be extracted so we don't have so much duplicated code that will need to be touched up in multiple places in the future?

Reply:
I guess it can. My approach was to touch as little external code as possible to avoid fucking something up.
I'll see what I can come up with.

Reply:
Understandable... but long-term maintenance is important.